contracts/Aavegotchi/facets/ItemsRolesRegistryFacet.sol

```diff
@@ -0,0 +1,253 @@
+        Commitment storage _commitment = _commitmentInfo.commitment;
+
+        address caller = _findCaller(_commitment.grantor, _roleAssignment.grantee, _commitment.tokenAddress);
+        if (_roleAssignment.expirationDate > block.timestamp && !_roleAssignment.revocable) {
+            // if role is not expired and is not revocable, only the grantee can revoke it
+            require(caller == _roleAssignment.grantee, "ItemsRolesRegistryFacet: role is not expired and is not revocable");
+        }
+
+        _unequipAllDelegatedWearables(_commitmentId, _commitment.tokenId);
+
+        emit RoleRevoked(_commitmentId, _role, _roleAssignment.grantee);
+        delete _commitmentInfo.roleAssignment;
+    }
+
+    function releaseTokens(
+        uint256 _commitmentId
+    ) external override onlyOwnerOrApproved(s.itemRolesCommitmentInfo[_commitmentId].commitment.grantor, s.itemRolesCommitmentInfo[_commitmentId].commitment.tokenAddress) {
+        ItemRolesInfo storage _commitmentInfo = s.itemRolesCommitmentInfo[_commitmentId];
+        Commitment memory _commitment = _commitmentInfo.commitment;
+        require(_commitment.tokenAmount > 0, "ItemsRolesRegistryFacet: commitment does not exist");
+        require(
+            _commitmentInfo.roleAssignment.expirationDate < block.timestamp || _commitmentInfo.roleAssignment.revocable,
+            "ItemsRolesRegistryFacet: token has an active role"
+        );
+
+        _unequipAllDelegatedWearables(_commitmentId, _commitment.tokenId); // If the item is equipped in some gotchi, it will be unequipped
+
+        delete _commitmentInfo.commitment;
+        delete _commitmentInfo.roleAssignment;
+
+        emit TokensReleased(_commitmentId);
+        _transferFrom(address(this), _commitment.grantor, _commitment.tokenAddress, _commitment.tokenId, _commitment.tokenAmount);
+    }
+
+    function setRoleApprovalForAll(address _tokenAddress, address _operator, bool _isApproved) external override {
+        s.itemsRoleApprovals[LibMeta.msgSender()][_tokenAddress][_operator] = _isApproved;
+        emit RoleApprovalForAll(_tokenAddress, _operator, _isApproved);
+    }
+
+    /** View Functions **/
+
+    function grantorOf(uint256 _commitmentId) override external view returns (address grantor_) {
+        grantor_ = s.itemRolesCommitmentInfo[_commitmentId].commitment.grantor;
+    }
+
+    function tokenAddressOf(uint256 _commitmentId) override external view returns (address tokenAddress_) {
+        tokenAddress_ = s.itemRolesCommitmentInfo[_commitmentId].commitment.tokenAddress;
+    }
+
+    function tokenIdOf(uint256 _commitmentId) override external view returns (uint256 tokenId_) {
+        tokenId_ = s.itemRolesCommitmentInfo[_commitmentId].commitment.tokenId;
+    }
+
+    function tokenAmountOf(uint256 _commitmentId) override external view returns (uint256 tokenAmount_) {
+        tokenAmount_ = s.itemRolesCommitmentInfo[_commitmentId].commitment.tokenAmount;
+    }
+
```

Comment:
i think it might be better to just have a single function that returns the full `ItemRolesInfo` instead of singluar functions unless we're just following ERC7589 standard

Reply:
Hi @Timidan 

This is a part of the ERC-7589. We considered using a single function, but the feedback we got from the review process is to avoid using structs on EIPs, and returning a single value copies less data to memory (e.g., if someone wants the tokenAmount, they would waste gas by copying the tokenId and other values to memory)